lib/service/service.go

```diff
@@ -1462,6 +1475,77 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	return nil
 }
 
+// readyzHandler serves the /readyz endpoint which can be used to fetch the
+// state of Teleport.
+func (process *TeleportProcess) readyzHandler(state *int64) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		switch atomic.LoadInt64(state) {
+		// 503
+		case stateDegraded:
+			roundtrip.ReplyJSON(w, http.StatusServiceUnavailable, map[string]interface{}{
+				"status": "teleport is in a degraded state, check logs for details",
+			})
+		// 400
+		case stateRecovering:
+			roundtrip.ReplyJSON(w, http.StatusBadRequest, map[string]interface{}{
+				"status": "teleport is recovering from a degraded state, check logs for details",
+			})
+		// 200
+		case stateOK:
+			roundtrip.ReplyJSON(w, http.StatusOK, map[string]interface{}{
+				"status": "ok",
+			})
+		}
+	}
+}
+
+// readyzMonitor monitors for events occuring within Teleport and updates
+// its internal state.
+func (process *TeleportProcess) readyzMonitor(state *int64) {
+	var recoveryTime time.Time
+
+	// Start loop to monitor for events that are used to update Teleport state.
+	eventCh := make(chan Event, 1024)
+	process.WaitForEvent(process.ExitContext(), TeleportReadyEvent, eventCh)
+	process.WaitForEvent(process.ExitContext(), TeleportDegradedEvent, eventCh)
+	process.WaitForEvent(process.ExitContext(), TeleportOKEvent, eventCh)
+
+	for {
+		select {
+		case e := <-eventCh:
+			switch e.Name {
+			// Ready event means Teleport has started successfully.
+			case TeleportReadyEvent:
```

Comment:
why not go one step further, and factor this even more? I was imagining something like this:

```
type StateMachine struct {
    recoveryTime time.Time
    currentState int64
}

func (f *StateMachine) Process(event Event) {
   atomic.StoreInt64(f.state, stateOK)
}

func (f *StateMachine) GetState() {
  atomic.LoadInt64(state, stateOK)
}
```

and then use this FSM